### PR TITLE
convert the HSE boundary conditions to C++

### PR DESCRIPTION
**Pull request by zingale**
## PR summary

As part of this change the *_ext_bc_type inputs parameters were changed from string to int.
This makes it easier to parse them.

## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
some work to do: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-11-12-001/index.html

File: Exec/gravity_tests/hse_convergence_general/inputs_1d.256.hse_test [hunk 1]
```diff
@@ -17,11 +17,15 @@ amr.n_cell           = 256
...
-castro.xl_ext_bc_type = "hse"
+castro.xl_ext_bc_type = 1
 castro.hse_interp_temp = 1
 castro.hse_reflect_vels = 1
 
+<<<<<<< HEAD
 castro.xr_ext_bc_type = "hse"
+=======
+castro.xr_ext_bc_type = 1
+>>>>>>> a1ff468c5c7afebde153f37ac41278680b8f1e45
```
Comment (maxpkatz) [inline Exec/gravity_tests/hse_convergence_general/inputs_1d.256.hse_test:28]: Merge conflict here

File: Exec/reacting_tests/reacting_bubble/inputs_2d [hunk 2]
```diff
@@ -17,7 +17,11 @@ amr.n_cell           = 384     640
 castro.lo_bc       =  0   1
 castro.hi_bc       =  0   2
 
+<<<<<<< HEAD
 castro.yl_ext_bc_type = "hse"
+=======
+castro.yl_ext_bc_type = 1
+>>>>>>> 3903438b28fa872edc9ac105744bac7c2512c2bb
```
Comment (maxpkatz) [inline Exec/reacting_tests/reacting_bubble/inputs_2d:24]: Merge conflict

File: Source/problems/Castro_bc_fill_nd.cpp [hunk 3]
```diff
@@ -121,6 +121,61 @@ void ca_statefill(Box const& bx, FArrayBox& data,
...
+    // corner where two (or three) faces want to do HSE, we may run
+    // into a situation that the data is not valid in the corner where
+    // we start the integration.  We'll abort, for now, if we run into
+    // this case.
+    //
+    // The future fix is to first call ext_fill on the ghost cells
+    // that are not corners and then call it a second time on just the
+    // corners.
+
+    std::cout << "about to check" << std::endl;
```
Comment (maxpkatz) [inline Source/problems/Castro_bc_fill_nd.cpp:135]: Leftover debug information

Comment (zingale) [reply Source/problems/Castro_bc_fill_nd.cpp]: fixed

**Comment by zingale** [comment]
tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-11-15-011/index.html

**Comment by zingale** [comment]
GPU flame_wave looks good: http://groot.astro.sunysb.edu/Castro/test-suite/gpu/2020-11-16-003/index.html

note: the 3-d has a compilation error from an earlier change.  I'll do a separate PR for that.

**Review by maxpkatz — APPROVED** [review]

**Pull request merged**